src/audio/sound_dose/sound_dose-generic.c

```diff
@@ -0,0 +1,291 @@
+			}
+		}
+
+		/* One of the buffers needs a wrap (or end of data), so check for wrap */
+		x0 += samples_without_wrap;
+		y0 += samples_without_wrap;
+		x0 = (x0 >= x_end) ? x0 - x_size : x0;
+		y0 = (y0 >= y_end) ? y0 - y_size : y0;
+
+		/* Update processed samples count for next loop iteration. */
+		samples -= samples_without_wrap;
+	}
+
+	/* Update the source and sink for bytes consumed and produced. Return success. */
+	source_release_data(source, bytes);
+	sink_commit_buffer(sink, bytes);
+
+	sound_dose_calculate_mel(mod, frames);
+	return 0;
+}
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S32LE || CONFIG_FORMAT_S32LE
+
+/**
+ * sound_dose_s32() - Process S32_LE or S24_4LE format.
+ * @mod: Pointer to module data.
+ * @source: Source for PCM samples data.
+ * @sink: Sink for PCM samples data.
+ * @frames: Number of audio data frames to process.
+ *
+ * Processing function for signed integer 32-bit PCM formats. The same
+ * function works for s24 and s32 formats since the samples values are
+ * not modified in computation. The audio samples in every frame are
+ * re-order to channels order defined in component data channel_map[].
+ *
+ * Return: Value zero for success, otherwise an error code.
+ */
+static int sound_dose_s32(const struct processing_module *mod,
+			  struct sof_source *source,
+			  struct sof_sink *sink,
+			  uint32_t frames)
+{
+	struct sound_dose_comp_data *cd = module_get_private_data(mod);
+	struct iir_state_df1 *iir;
+	int32_t sample;
+	int32_t const *x0, *x, *x_start, *x_end;
+	int32_t *y0, *y, *y_start, *y_end;
+	int32_t weighted;
+	int x_size, y_size;
+	int source_samples_without_wrap;
+	int samples_without_wrap;
+	int samples = frames * cd->channels;
+	int bytes = frames * cd->frame_bytes;
+	int ret;
+	int ch;
+	int i;
+	const int channels = cd->channels;
+
+	/* Get pointer to source data in circular buffer, get buffer start and size to
+	 * check for wrap. The size in bytes is converted to number of s16 samples to
+	 * control the samples process loop. If the number of bytes requested is not
+	 * possible, an error is returned.
+	 */
+	ret = source_get_data_s32(source, bytes, &x0, &x_start, &x_size);
+	if (ret)
+		return ret;
+
+	/* Similarly get pointer to sink data in circular buffer, buffer start and size. */
+	ret = sink_get_buffer_s32(sink, bytes, &y0, &y_start, &y_size);
+	if (ret)
+		return ret;
+
+	/* Set helper pointers to buffer end for wrap check. Then loop until all
+	 * samples are processed.
+	 */
+	x_end = x_start + x_size;
+	y_end = y_start + y_size;
+	while (samples) {
+		/* Find out samples to process before first wrap or end of data. */
+		source_samples_without_wrap = x_end - x0;
+		samples_without_wrap = y_end - y0;
+		samples_without_wrap = MIN(samples_without_wrap, source_samples_without_wrap);
+		samples_without_wrap = MIN(samples_without_wrap, samples);
+		for (ch = 0; ch < cd->channels; ch++) {
+			iir = &cd->iir[ch];
+			x = x0++;
+			y = y0++;
```

Comment:
is this correct? `x0` and `y0` are incremented here for each channel and then in lines 239-242 they're used for wrapping?